File: src/collectors/protocol.mjs
export const SDK_VERSION = '1';
export const MAX_ENVELOPE_BYTES = 262_144;

const COLLECTOR_ID = /^[A-Za-z0-9][A-Za-z0-9_-]*(?:\.[A-Za-z0-9][A-Za-z0-9_-]*){2,}$/;
const CAPABILITIES = new Set([
  'process.lifecycle',
  'process.metadata',
  'file.metadata',
  'collector.diagnostics',
]);

function protocolError() {
  const error = new Error('COLLECTOR_PROTOCOL_ERROR');
  error.code = 'COLLECTOR_PROTOCOL_ERROR';
  return error;
}

function sequenceError() {
  const error = new Error('COLLECTOR_SEQUENCE_ERROR');
  error.code = 'COLLECTOR_SEQUENCE_ERROR';
  return error;
}

function isRecord(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function validCollectorId(value) {
  return typeof value === 'string' && COLLECTOR_ID.test(value);
}

function encodedBytes(value) {
  try {
    return Buffer.byteLength(JSON.stringify(value), 'utf8');
  } catch {
    throw protocolError();
  }
}

export function validateManifest(value) {
  if (!isRecord(value)
    || value.sdkVersion !== SDK_VERSION
    || !validCollectorId(value.id)
    || typeof value.name !== 'string' || !value.name.trim()
    || typeof value.version !== 'string' || !value.version.trim()
    || !Array.isArray(value.platforms) || value.platforms.length === 0
    || !Array.isArray(value.capabilities)
    || value.capabilities.some((capability) => !CAPABILITIES.has(capability))
    || typeof value.requiresExplicitPaths !== 'boolean'
    || typeof value.requiresTargetLaunch !== 'boolean'
    || value.contentCapture !== 'unsupported') {
    throw protocolError();
  }
  return value;
}

export function validateEnvelope(value, previousSequence = null) {
  if (!isRecord(value)
    || value.sdkVersion !== SDK_VERSION
    || !validCollectorId(value.collectorId)
    || typeof value.instanceId !== 'string' || !value.instanceId.trim()
    || !Number.isSafeInteger(value.sequence) || value.sequence < 0
    || !['event', 'diagnostic', 'heartbeat', 'completed'].includes(value.kind)) {
    throw protocolError();
  }

  if (previousSequence !== null && value.sequence <= previousSequence) {
    throw sequenceError();
  }

  const hasEvent = value.event !== undefined && value.event !== null;
  const hasDiagnostic = value.diagnostic !== undefined && value.diagnostic !== null;
  const shapeValid = value.kind === 'event'
    ? hasEvent && !hasDiagnostic
    : value.kind === 'diagnostic'
      ? !hasEvent && hasDiagnostic
      : !hasEvent && !hasDiagnostic;

  if (!shapeValid || encodedBytes(value) > MAX_ENVELOPE_BYTES) {
    throw protocolError();
  }
  return value;
}

File: crates/harnesscope-collector-sdk/tests/protocol.rs
use harnesscope_collector_sdk::{
    CollectorCapability, CollectorEnvelope, CollectorEnvelopeKind, CollectorManifest,
    CollectorProtocolError, COLLECTOR_SDK_VERSION, MAX_ENVELOPE_BYTES, validate_envelope,
    validate_manifest,
};
use serde_json::json;

fn fixture_manifest() -> CollectorManifest {
    CollectorManifest {
        sdk_version: "1".into(),
        id: "harnesscope.synthetic.collector".into(),
        name: "Synthetic Collector".into(),
        version: "0.4.0".into(),
        platforms: vec!["linux".into(), "macos".into()],
        capabilities: vec![
            CollectorCapability::ProcessLifecycle,
            CollectorCapability::ProcessMetadata,
            CollectorCapability::FileMetadata,
            CollectorCapability::CollectorDiagnostics,
        ],
        requires_explicit_paths: true,
        requires_target_launch: true,
        content_capture: "unsupported".into(),
    }
}

fn fixture_envelope(sequence: u64) -> CollectorEnvelope {
    CollectorEnvelope {
        sdk_version: "1".into(),
        collector_id: "harnesscope.synthetic.collector".into(),
        instance_id: "instance-1".into(),
        sequence,
        kind: CollectorEnvelopeKind::Event,
        event: Some(json!({
            "source": "collector",
            "kind": "ProcessStarted",
            "correlationId": "pid:1234",
            "data": { "pid": 1234 }
        })),
        diagnostic: None,
    }
}

#[test]
fn exposes_exact_v1_constants() {
    assert_eq!(COLLECTOR_SDK_VERSION, "1");
    assert_eq!(MAX_ENVELOPE_BYTES, 262_144);
}

#[test]
fn validates_manifest_contract() {
    let manifest = fixture_manifest();
    validate_manifest(&manifest).unwrap();

    let mut bad_version = manifest.clone();
    bad_version.sdk_version = "2".into();
    assert!(matches!(
        validate_manifest(&bad_version),
        Err(CollectorProtocolError::Protocol)
    ));

    let mut bad_id = manifest;
    bad_id.id = "not-a-reverse-dns-id".into();
    assert!(matches!(
        validate_manifest(&bad_id),
        Err(CollectorProtocolError::Protocol)
    ));
}

#[test]
fn rejects_duplicate_or_out_of_order_sequence() {
    let first = fixture_envelope(7);
    validate_envelope(None, &first).unwrap();
    assert!(matches!(
        validate_envelope(Some(7), &fixture_envelope(7)),
        Err(CollectorProtocolError::Sequence)
    ));
    assert!(matches!(
        validate_envelope(Some(7), &fixture_envelope(6)),
        Err(CollectorProtocolError::Sequence)
    ));
}

#[test]
fn rejects_shape_mismatches_and_oversize_envelopes() {
    let mut mismatch = fixture_envelope(1);
    mismatch.diagnostic = Some(harnesscope_collector_sdk::CollectorDiagnostic {
        code: "BAD".into(),
        message: "x".into(),
        data: None,
    });
    assert!(matches!(
        validate_envelope(None, &mismatch),
        Err(CollectorProtocolError::Protocol)
    ));

    let mut heartbeat = fixture_envelope(2);
    heartbeat.kind = CollectorEnvelopeKind::Heartbeat;
    assert!(matches!(
        validate_envelope(None, &heartbeat),
        Err(CollectorProtocolError::Protocol)
    ));

    let mut oversized = fixture_envelope(3);
    oversized.event = Some(json!({
        "source": "collector",
        "kind": "Unknown",
        "data": { "payload": "x".repeat(MAX_ENVELOPE_BYTES) }
    }));
    assert!(matches!(
        validate_envelope(None, &oversized),
        Err(CollectorProtocolError::Protocol)
    ));
}
